### PR TITLE
Replace Config class with ConfigReader class

### PR DESCRIPTION
**Pull request by theseanything**
This PR aims are replacing the Config class with a ConfigReader class. This is apart of an effort to simplify constructing Stack and Environment objects, addressing issue #5. 

**Currently** Sceptre's main entry point when using it as Python module is to create an environment object using the `Environment(sceptre_dir, environment_path, options=None)` constructor. The Environment class is responsible for reading the relevant config from a Sceptre project folder and creating Stack objects. Additionally Config class is a subclass of the built-in dict with custom class attributes - which make testing incredible difficult and confusing on how to access relevant properties (i.e. is it a key or a attribute - `config["region"]` vs `config.sceptre_dir`). 

**_The new_** ConfigReader class aims to fix these issues - by encapsulating functions related to gathering/parsing Sceptre project folders and reading in config (instead of the Environment class). -
- Config read in from a file is now represented by a "vanilla" python dict - simplifying where properties are located. 
- The `construct_stack` and `construct_environment` methods now provide the ability to generate Stack or Environment object respectively using config files. 
- This cleans up the constructor of the Environment class and adds the ability of generating Stack objects without having to make a Environment object first.

This API change include referencing stacks and environment by full relative paths - laying the foundation to solve issue #62.

-----------------

* [x] Wrote [good commit messages][1].
* [x] [Squashed related commits together][2] after the changes have been approved.
* [x] Commit message starts with `[Resolve #issue-number]` (if a related issue exists).
* [x] Added unit tests.
* [x] Added integration tests (if applicable).
* [x] All unit tests (`make test`) are passing.
* [x] Used the same coding conventions as the rest of the project.
* [x] The new code doesn't generate flake8 (`make lint`) offenses.
* [ ] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences. - Sort of, difficult to do in stages.

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

### CODE REVIEW

File: sceptre/config_reader.py [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    produces Stack or Environment objects. Responsible for loading Resolvers
+    and Hook classes and adding them as constructors to the PyYAML parser.
+
+    :param sceptre_dir: The absolute path to the Sceptre directory.
+    :type sceptre_dir: str
+    """
+
+    def __init__(self, sceptre_dir, variables=None):
+        self.logger = logging.getLogger(__name__)
+        self._deferred_constructors = []
```
Comment (theseanything) [inline sceptre/config_reader.py]: remove

File: sceptre/config_reader.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+                protected=config.get("protect", False),
+                tags=config.get("stack_tags", {}),
+                external_name=config.get("external_name")
+            )
+
+            self._construct_nodes(config, stack)
+            del self.templating_vars["environment_config"]
+            return stack
+
+    def construct_stack(self, rel_path):
```
Comment (theseanything) [inline sceptre/config_reader.py]: rel_path not clear

File: sceptre/environment.py [hunk 3]
```diff
@@ -42,77 +36,27 @@ class Environment(object):
...
         self.logger = logging.getLogger(__name__)
+        self.path = path
 
-        self.sceptre_dir = sceptre_dir
-        self.path = self._validate_path(environment_path)
-        self._options = {} if options is None else options
-
-        self._is_leaf = None
+        self.stacks = []
+        self.environments = []
```
Comment (ngfgrant) [inline sceptre/environment.py]: Potentially rename to `sub_environments`

File: integration-tests/steps/change_sets.py [hunk 4]
```diff
@@ -49,9 +48,8 @@ def step_impl(context, stack_name):
 
 @when('the user creates change set "{change_set_name}" for stack "{stack_name}"')
 def step_impl(context, change_set_name, stack_name):
-    environment_name, basename = os.path.split(stack_name)
-    env = Environment(context.sceptre_dir, environment_name)
-    stack = env.stacks[basename]
+    config_reader = ConfigReader(context.sceptre_dir)
+    stack = config_reader.construct_stack(stack_name + ".yaml")
```
[the remark this answers is not in the document]
Comment (theseanything) [reply integration-tests/steps/change_sets.py]: This is just for the stack config which can only be .yaml at the moment - but yeah possible improvement at a later date.

File: sceptre/cli.py [hunk 5]
```diff
@@ -735,24 +735,43 @@ def _create_config_file(config_dir, path, defaults={}):
...
     """
     Initialises and returns a sceptre.environment.Environment().
 
     :param sceptre_dir: The absolute path to the Sceptre directory.
     :type project dir: str
-    :param environment_path: The name of the environment.
-    :type environment_path: str
-    :param options: A dict of key-value pairs to update self.config with.
-    :type debug: dict
+    :param environment: The name of the environment.
```
[the remark this answers is not in the document]
Comment (theseanything) [reply sceptre/cli.py]: We've used full stops everywhere else, so will just stick with it for now to be consistent.

File: sceptre/config_reader.py (hunk 2 above)
Comment (theseanything) [reply sceptre/config_reader.py]: Clarified

**Pull request merged**